Run filters without inputs first in OrcSelectiveStreamReader

sayhar opened the pull request:
Filters without any inputs are usually less expensive and
more productive than other filters, so it is more efficient
to run them first.

```
== NO RELEASE NOTE ==
```

sayhar commented:
This is a stripped down version of the first commit of https://github.com/prestodb/presto/pull/13449/

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -428,6 +439,27 @@ private void initializePositions(int batchSize)
+            }
+        }
+
+        // at this point outputPositions are relative to page, e.g. they are indices into positions array
```
Since no-input functions run first, this translation is not needed.

sayhar replied:
That's right!

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -496,7 +528,9 @@ private void initializeOutputPositions(int positionCount)
         }
-
+    }
+    private void initalizeErrors(int positionCount)
```
add empty line before the method

sayhar replied:
Will do!

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -165,7 +166,12 @@ public OrcSelectiveRecordReader(
         this.columnTypes = includedColumns.entrySet().stream().collect(toImmutableMap(entry -> zeroBasedIndices.get(entry.getKey()), Map.Entry::getValue));
-        this.filterFunctions = filterFunctions;
+        this.filterFunctionsWithoutInputs = filterFunctions.stream()
+                .filter(x -> x.getInputChannels().length == 0)
```
Thanks!

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -165,7 +166,12 @@ public OrcSelectiveRecordReader(
+                .filter(x -> x.getInputChannels().length == 0)
+                .collect(toImmutableList());
+        this.filterFunctionsWithInputs = filterFunctions.stream()
+                .filter(x -> x.getInputChannels().length > 0)
```
Okay!

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -341,10 +347,15 @@ public Page getNextPage()
         int[] positionsToRead = this.positions;
         int positionCount = batchSize;
-        boolean filterFunctionsApplied = filterFunctions.isEmpty();
+        boolean filterFunctionsApplied = filterFunctionsWithInputs.isEmpty();
```
Okay!

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -165,7 +166,12 @@ public OrcSelectiveRecordReader(
         this.outputColumns = outputColumns.stream().map(zeroBasedIndices::get).collect(toImmutableList());
         this.columnTypes = includedColumns.entrySet().stream().collect(toImmutableMap(entry -> zeroBasedIndices.get(entry.getKey()), Map.Entry::getValue));
-        this.filterFunctions = filterFunctions;
+        this.filterFunctionsWithoutInputs = filterFunctions.stream()
```
Good point!

sayhar commented:
Not sure this works. Running tests locally is going *very slowly* for me so I'm pushing this update to see if it passes the tests. 

Also -- we caught a bug in the proposal where the no-input filterfunction (or the query itself) has positioncount = 0 right away. so this hopefully fixes that. Not sure it's as elegant as it could be

sayhar commented:
This new code switches filterFunctionsWithNoInputs to an Optional<Function>, like Masha suggested. 

In doing so, I switched a use of outputPositions to positions. Not sure that's right. Would appreciate a sanity check on that.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -165,7 +166,18 @@ public OrcSelectiveRecordReader(
-        this.filterFunctions = filterFunctions;
+        List<FilterFunction> filterFunctionsWithoutInputsList = filterFunctions.stream()
+                .filter(function -> function.getInputChannels().length == 0)
+                .limit(1)
```
let's make sure there will be an error if there is more than one such function:

```
this.filterFunctionWithoutInput = getFilterFunctionWithoutInputs(filterFunctions);


    private static Optional<FilterFunction> getFilterFunctionWithoutInputs(List<FilterFunction> filterFunctions)
    {
        List<FilterFunction> functions = filterFunctions.stream()
                .filter(function -> function.getInputChannels().length == 0)
                .collect(toImmutableList());
        if (functions.isEmpty()) {
            return Optional.empty();
        }

        return Optional.of(Iterables.getOnlyElement(functions));
    }
```

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -341,10 +353,20 @@ public Page getNextPage()
+        }
+        boolean filterFunctionsApplied = filterFunctionsWithInputs.isEmpty();
+
+        if (positionCount == 0) {
```
- move this inside the `if (filterFunctionWithoutInput.isPresent())` block
- call `batchRead(batchSize);` before returning

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -428,6 +450,14 @@ private void initializePositions(int batchSize)
         }
     }
 
+    private int applyFilterFunctionsWithNoInputs(int[] positions, int positionCount)
```
applyFilterFunctionsWithNoInputs -> applyFilterFunctionWithNoInputs (plural to singular)

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -428,6 +450,14 @@ private void initializePositions(int batchSize)
+    {
+        initializeOutputPositions(positionCount);
+        Page page = new Page(positionCount);
+        positionCount = filterFunctionWithoutInput.get().filter(page, positions, positionCount, errors);
```
- avoid re-assigning input parameters; 

`return filterFunctionWithoutInput.get().filter(page, positions, positionCount, errors);`

sayhar replied:
When you say positions, do you mean we can use positions instead of outputPositions?

sayhar commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -131,9 +132,32 @@ private static QueryRunner createQueryRunner()
                         "   IF (orderkey % 49 = 0, NULL, ARRAY[CAST(comment AS CHAR(4)), CAST(comment AS CHAR(3)), CAST(SUBSTR(comment,length(comment) - 4) AS CHAR(4))]) AS char_array, " +
                         "   IF (orderkey % 49 = 0, NULL, ARRAY[comment, comment]) AS varchar_array " +
                         "FROM lineitem");
+
```
I'll remove these new linebreaks, no worries

sayhar commented:
@mbasmanova as I messaged you privately, since we last spoke:

1. (Minor): assertQuerySucceeds existed in TestTpcds, and was now a private method overriding a protected one. So I just changed it to protected. (Optionally, I could have changed the name instead)
2. (Less minor): I changed round() -> floor() in the new test code. This removed test flakiness.

mbasmanova commented on presto-tpcds/src/test/java/com/facebook/presto/tpcds/TestTpcds.java:
```diff
@@ -72,7 +72,7 @@ public void testLargeInWithShortDecimal()
     }
 
-    private void assertQuerySucceeds(String sql)
+    protected void assertQuerySucceeds(String sql)
```
@sayhar just delete this method and use the one from the base class
